### DISCUSSION TITLE
Welcome to svelte-range-slider-pips Discussions!

### DISCUSSION THREAD

**Discussion by simeydotme** [simeydotme/svelte-range-slider-pips#17] [General]
## 👋 Discuss!
  Enabling this feature to move questions / ideas / support for the Component (either Svelte, or Vanilla, or Web Components) and Testing out of the Issues area. Please feel free to;
  * Ask questions about functionality / how to.
  * Share ideas for improvements / new features.
  * Engage with other users and share use-cases (like a showcase).
  * Welcome others and be open-minded. Remember that this is a community we
  build together 💪.

Let's see how this goes! :)

**Comment by amandakoster** [comment]
Hello, thank you for such a great component with equally great documentation! I am wondering how you printed the range values in your Example page: https://simeydotme.github.io/svelte-range-slider-pips/#svelte-range-slider--pips.  Even better where is the code for this pageL Examples.svelte so I can figure it out myself?

**Reply by simeydotme** [reply]
https://github.com/simeydotme/svelte-range-slider-pips/blob/gh-pages/src/

^^  So the Docs are on the `gh-pages` branch, so glad you're interested to figure out ! :)
The main file is a `.svx` which is like a **markdown** file with code interspersed. And the code for the values-printing is in the `Example.svelte`.

Let me know if you struggle at all!

**Reply by simeydotme** [reply]
I realise the component structure in the `gh-pages` branch is a little muddy, so if you want a shortcut here's an example of values being displayed;

https://svelte.dev/repl/f7bcc48c34b745c2b6b0d622410b2ec6?version=3.49.0

- If you're using the slider with Vanilla / jQuery, you might need to use the `onchange` event [(example on codepen)](https://codepen.io/simeydotme/pen/KKNJdbK) 
- If you're using it with React/Vue, I'm not sure quite how you can bind reactive values; so if you figure it out, please share back here! :)

**Reply by amandakoster** [reply]
Found it, thank you!

**Comment by amandakoster** [comment]
Hello, I am running into an issue. I am trying to use this component for a date range slider, so the user can choose a range of dates within a max range (and then search those dates).  My issue is that once I format the unix time stamp to a human readable date, the slider crashes because it does not know how to read a date (ex., 1/1/2010) vs. a number  (123456)

Example:
https://svelte.dev/repl/9be23a7f10aa417d8b530ee8aa03cbae?version=3.43.1

**Reply by simeydotme** [reply]
hah... oh man ... I know it's purely problem-bias, but there's so many funny/un-anticipated use-cases for these Sliders 🙈 
Let me check if I can help out :)

**Reply by simeydotme** [reply]
Hi @amandakoster  ... here is something I'm sure you can extract value / inspiration from :)

https://svelte.dev/repl/76c7d258986a4f7b975f9cb5d62a10c6?version=3.43.1
^^ It has a ton of date formatting,.I'm using the `step`/`pipstep` properties as well as the `formatter` methods, to get the slider UI to a decent usable state... but I imagine you'll need to tweak for your own use-case.

Thanks for your interest! if it's something you endorse then it would be great if you share the component, help out with code/docs, or support in any way you can! cheers

**Reply by amandakoster** [reply]
This is awesome! Thank you so much.  However. I am getting this error after using the code in the Repl: 
```
'values' prop should be an Array (https://github.com/simeydotme/svelte-range-slider-pips#slider-props)

Uncaught (in promise) RangeError: Invalid array length
    at create_if_block_4 (RangePips.svelte:195:12)
    at create_fragment$1 (RangePips.svelte:194:41)
    at init (index.mjs:1891:37)
    at new RangePips (RangePips.svelte:53:48)
    at create_if_block$1 (RangeSlider.svelte:832:58)
    at create_fragment$2 (RangeSlider.svelte:834:12)
    at init (index.mjs:1891:37)
    at new RangeSlider (RangeSlider.svelte:151:91)
    at create_fragment$3 (Slider.svelte:19:21)
    at init (index.mjs:1891:37)
```

This is my code:

```
<script>
  import Slider from '../../Slider.svelte';  // imports the Slider with props

  export const oneDay = 86400000;
  export const oneYear = 31560000000;

  // Unix timestamp range
  export let minUnix = 1262332800; // 1/1/2010
  export let maxUnix = 2209017600; // 1/1/2040

  export const getMilliseconds = ms => ms * 1000;
  export let minMs = getMilliseconds(minUnix);
  export let maxMs = getMilliseconds(maxUnix);

  export const minDateObject = new Date(minMs);
  export const maxDateObject = new Date(maxMs);

  // MIN
  export let minMonth = minDateObject.getUTCMonth() + 1; // months from 1-12
  export let minDay = minDateObject.getUTCDate();
  export let minYear = minDateObject.getUTCFullYear();

  // MAX
  export let maxMonth = maxDateObject.getUTCMonth() + 1; // months from 1-12
  export let maxDay = maxDateObject.getUTCDate();
  export let maxYear = maxDateObject.getUTCFullYear();

  export const minDate = `${minDay}/${minMonth}/${minYear}`;
  export const maxDate = `${maxDay}/${maxMonth}/${maxYear}`;

  export const format = v => {
    let date = new Date(v).setDate(1);
    // have to set day to 1, as dates and sliders are not exactly compatible as
    // years and months have variable time in them.
    return new Intl.DateTimeFormat('en-US').format(date); // get mm/dd/yyyy
    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat
  };

  export const formatDate = v => {
    const d = format(v);
    return `${d.split('/')[0].padStart(2, '0')}/${d.split('/')[2]}`; // get mm/yyyy
  };

  export const formatYear = v => {
    const d = format(v);
    return d.split('/')[2]; // get the year
  };

  export let values = [minMs, maxMs];
  console.log('typeof values', typeof values);
  $: formattedValues = values.map(v => format(v));
  // ^^ reactive variable
</script>

<h3>range: {formatDate(minMs)} ~ {formatDate(maxMs)}</h3>
<h3>formattedValues: {formattedValues}</h3>

<Slider
  bind:values
  min={minMs}
  max={maxMs}
  step={oneYear / 12}
  pipstep={60}
  formatter={formatYear}
  handleFormatter={formatDate}
  range
  float
  pips
  all="label"
/>
```
this logs as an object:
`console.log('typeof values', typeof values);`
`typeof values object`

**Reply by amandakoster** [reply]
Also, do we need to add everything in `RangePips.svelte` and `RangeSlider.svelte` from the Repl into my code or does that come with the package?

**Reply by simeydotme** [reply]
hmm, sorry @amandakoster that is a issue with your setup/environment ... there's a regression somewhere in your code. 

I'd hazard a guess it's to do with all the `export let` and `export const` ... I doubt you actually need to expose all those props, and it looks like a code-smell. And if you're doing anything with the `values` prop inside `Slider.svelte`, it could also be the cause.

If you get stuck, you could try the Svelte Discord for some troubleshooting :)

**Reply by simeydotme** [reply]
> Also, do we need to add everything in RangePips.svelte and RangeSlider.svelte from the Repl into my code or does that come with the package?

If you use `import RangeSlider from "./RangeSlider.svelte";` and have done `npm install --save-dev svelte-range-slider-pips` then I think you should be fine :) (ie: https://svelte.dev/repl/030797781fd64ad88302d1343f5b2c43?version=3.49.0) .. but again I'm not sure what your environment/setup is ... this is just the regular NPM / Svelte / Package method

**Reply by amandakoster** [reply]
This is a huge help! Thank you. I will try those out!

**Reply by amandakoster** [reply]
IT WORKED!!!!! You are amazing!!! Thank you so much!

**Reply by simeydotme** [reply]
yay!

**Reply by simeydotme** [reply]
@amandakoster may I ask; how/where did you hear about this component? and would you consider adding a ⭐ ... it helps gain exposure :)
